File: development/NET/samples/client/ClientDLL/VB/OpcProject/OpcClient.vb
Imports System
Imports Microsoft.VisualBasic
Imports Softing.OPCToolbox.Client
Imports Softing.OPCToolbox

Namespace OpcClient
    Public Class OpcClient

#Region "Fields"
        Dim m_traceGroupUser As EnumTraceGroup
#End Region

#Region "Constructor"
        '-----------------

        Public Sub New()
            m_traceGroupUser = [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.USER)
        End Sub

        '--
#End Region

#Region "Private Members"
        '---------------------

#Region "Private Attributes"
        '------------------------

        Private m_daSession As DaSession = Nothing
        Private m_daSubscription As DaSubscription = Nothing
        Private m_daItem As DaItem = Nothing
        Private m_aeSession As AeSession = Nothing
        Private m_aeSubscription As AeSubscription = Nothing
        '--
#End Region

#Region "Private Methods"
        '---------------------

        Private Function GetStateToString(ByVal state As EnumConditionState) As String
            Dim stateToString As String = String.Empty

            If (state And EnumConditionState.ACTIVE) = EnumConditionState.ACTIVE Then
                stateToString &= " ACT"
            End If
            If (state And EnumConditionState.ENABLED) = EnumConditionState.ENABLED Then
                stateToString &= " ENA"
            End If
            If (state And EnumConditionState.ACKED) = EnumConditionState.ACKED Then
                stateToString &= " ACK"
            End If
            If state = 0 Then
                stateToString &= " DIS"
            End If
            Return stateToString
        End Function

        '--
#End Region

        '--
#End Region

#Region "Public Methods"
        '---------------------

        Public Function GetApplication() As Application
            Return Application.Instance
        End Function

        Public Function Initialize() As Integer
            GetApplication().VersionOtb = 447
            Dim result As Integer = GetApplication().Initialize()
            Dim traceGroupClient As EnumTraceGroup
            Dim traceGroupAll As EnumTraceGroup
            traceGroupClient = [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.CLIENT)
            traceGroupAll = [Enum].ToObject(GetType(EnumTraceGroup), EnumTraceGroup.ALL)
            GetApplication().EnableTracing(traceGroupAll, traceGroupAll, traceGroupClient, traceGroupClient, "Trace.txt", Convert.ToUInt32(1000000), _
         Convert.ToUInt32(2))

            Return result
        End Function

        Public Function ProcessCommandLine(ByVal commandLine As String) As Integer
            Return Application.Instance.ProcessCommandLine(commandLine)
        End Function


        Public Sub Terminate()
            GetApplication().Terminate()
            m_daSession = Nothing
            m_daSubscription = Nothing
            m_daItem = Nothing
            m_aeSession = Nothing
            m_aeSubscription = Nothing
        End Sub


        Public Function GetConditionState() As String
            If m_aeSession Is Nothing Then
                Return ""
            End If

            Dim message As String = [String].Empty

            Try
                Dim conditionStateToString As String = [String].Empty
                Dim conditionState As AeConditionState = Nothing

                Dim result As Integer = m_aeSession.GetAeConditionState("computer.clock.time slot 1", "between", Nothing, conditionState, Nothing)

                If ResultCode.SUCCEEDED(result) Then

                    message = "The condition state is: " & Chr(10) & ""
                    message &= " Source Path: computer.clock.time slot 1 " & Chr(10) & ""
                    message &= " Condition Name: between " & Chr(10) & ""
                    message &= " State: " + GetStateToString(conditionState.State) + "" & Chr(10) & ""
                    message &= " Quality: " + conditionState.Quality.ToString() + "" & Chr(10) & ""
                    message &= " Active Time: " + conditionState.ConditionActiveTime + "" & Chr(10) & ""
                    message &= " Inactive Time: " + conditionState.ConditionInactiveTime + "" & Chr(10) & ""
                    message &= " AcknowledgeTime: " + conditionState.ConditionAckTime + "" & Chr(10) & ""
                    message &= " AcknowledgerComment: " + conditionState.AcknowledgerComment + "" & Chr(10) & ""
                    message &= " AcknowledgerID: " + conditionState.AcknowledgerId + "" & Chr(10) & ""
                    message &= " Active subcondition time: " + conditionState.SubConditionActiveTime + "" & Chr(10) & ""
                    message &= " Active subcondition name: " + conditionState.ActiveSubConditionName + "" & Chr(10) & ""
                    message &= " Active subcondition definition: " + conditionState.ActiveSubConditionDefinition + "" & Chr(10) & ""
                    message &= " Active subcondition description: " + conditionState.ActiveSubConditionDescription + "" & Chr(10) & ""
                    message &= " Active subcondition severity: " + conditionState.ActiveSubConditionSeverity.ToString() + "" & Chr(10) & ""
                    message &= " Number of subconditions: " + conditionState.SubConditionsNames.Length + "" & Chr(10) & ""
                    Dim i As Integer = 0
                    While i < conditionState.SubConditionsNames.Length
                        message &= "  Subcondition name: " + conditionState.SubConditionsNames(i) + "" & Chr(10) & ""
                        message &= "  Subcondition definition: " + conditionState.SubConditionsDefinitions(i) + "" & Chr(10) & ""
                        message &= "  Subcondition description: " + conditionState.SubConditionsDescriptions(i) + "" & Chr(10) & ""
                        message &= "  Subcondition severity: " + conditionState.SubConditionsSeverities(i).ToString() + "" & Chr(10) & ""
                        System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
                    End While
                Else
                    message = "Get condition state failed!" & Chr(10) & ""
                End If
            Catch exc As Exception
                GetApplication().Trace(EnumTraceLevel.ERR, m_traceGroupUser, "OpcClient::GetConditionState", exc.ToString())
            End Try
            Return message
        End Function

        Public Function InitializeAeObjects() As Integer
            Dim connectResult As Integer = EnumResultCode.E_FAIL
            Try

                m_aeSession = New AeSession("opcae:///Softing.OPCToolboxDemo_ServerAE.1/{2E565243-B238-11D3-842D-0008C779D775}")
                m_aeSubscription = New AeSubscription(m_aeSession)

                connectResult = m_aeSession.Connect(True, True, Nothing)

                m_aeSubscription.RefreshAeConditions(Nothing)
            Catch exc As Exception
                GetApplication().Trace(EnumTraceLevel.ERR, m_traceGroupUser, "OpcClient::InitializeAeObjects", exc.ToString())
            End Try

            Return connectResult
        End Function
        Public Function InitializeDaObjects() As Integer
            Dim connectResult As Integer = EnumResultCode.E_FAIL

            Try
                m_daSession = New DaSession("opcda:///Softing.OPCToolboxDemo_ServerDA.1/{2E565242-B238-11D3-842D-0008C779D775}")
                m_daSubscription = New DaSubscription(Convert.ToUInt32(1000), m_daSession)
                m_daItem = New DaItem("maths.sin", m_daSubscription)

                connectResult = m_daSession.Connect(True, False, Nothing)
            Catch exc As Exception
                GetApplication().Trace(EnumTraceLevel.ERR, m_traceGroupUser, "OpcClient::InitializaDaObjects", exc.ToString())
            End Try

            Return connectResult
        End Function

        Public Function ReadItem() As String
            Dim message As String = [String].Empty

            Try
                Dim readCount As UInt32 = Convert.ToUInt32(0)
                Dim itemList As DaItem() = New DaItem(Convert.ToInt32(readCount)) {}
                itemList(0) = m_daItem

                Dim values As ValueQT() = Nothing
                Dim results As Integer() = Nothing

                If ResultCode.SUCCEEDED(m_daSubscription.Read(Convert.ToUInt32(0), itemList, values, results, Nothing)) Then
                    message &= " " & Chr(10) & "Read item synchronously using subscription " & Chr(10) & ""

                    Dim i As Integer = 0
                    While i < values.Length
                        If ResultCode.SUCCEEDED(results(i)) Then
                            message &= " " + itemList(i).Id + " - "
                            message += values(i).ToString() + "" & Chr(10) & "" & Chr(10) & ""
                        Else
                            message &= "Read failed for item " + itemList(i).Id + "" & Chr(10) & "" & Chr(10) & ""
                        End If
                        System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
                    End While
                Else
                    message &= " Subscription synchronous read failed!" + "" & Chr(10) & "" & Chr(10) & ""
                End If
            Catch exc As Exception
                GetApplication().Trace(EnumTraceLevel.ERR, m_traceGroupUser, "OpcClient::ReadItem", exc.ToString())
            End Try
            Return message
        End Function


        Public Sub Trace(ByVal traceLevel As EnumTraceLevel, ByVal traceGroup As EnumTraceGroup, ByVal objectID As String, ByVal message As String)
            Application.Instance.Trace(traceLevel, traceGroup, objectID, message)
        End Sub

        '--
#End Region

#Region "Public Properties"
        '-----------------------

        Public Property ServiceName() As String
            Get
                Return Application.Instance.ServiceName
            End Get
            Set(ByVal Value As String)
                Application.Instance.ServiceName = Value
            End Set
        End Property

        '--
#End Region

    End Class

End Namespace

File: development/NET/samples/client/ClientDLL/CS/OpcProject/OpcClient.cs
using System;
using Softing.OPCToolbox.Client;
using Softing.OPCToolbox;

namespace TestDll
{
	public class OpcClient
	{
	
		#region Constructor
		//-----------------

		public OpcClient(){}

		//--
		#endregion

		#region Private Members
		//---------------------

		#region Private Attributes
		//------------------------

		private DaSession m_daSession = null;
		private DaSubscription m_daSubscription = null;
		private DaItem m_daItem = null;
		private AeSession m_aeSession = null;
		private AeSubscription m_aeSubscription = null;
		//--
		#endregion

		#region Private Methods
		//---------------------

		private string GetStateToString(EnumConditionState state)
		{
			string stateToString = string.Empty;

			if((state & EnumConditionState.ACTIVE) == EnumConditionState.ACTIVE)
			{
				stateToString += " ACT";
			}	//	end if
			if((state & EnumConditionState.ENABLED) == EnumConditionState.ENABLED)
			{
				stateToString += " ENA";
			}	//	end if
			if((state & EnumConditionState.ACKED) == EnumConditionState.ACKED)
			{
				stateToString += " ACK";
			}	//	end if
			if(state == 0)
			{
				stateToString += " DIS";
			}	//	end if
			return stateToString;
		}	//	end StateToString

		//--
		#endregion

		//--
		#endregion
		
		#region Public Methods
		//---------------------

		public Application GetApplication()
		{
			return Application.Instance;
		}	//	end GetApplication

		public int Initialize()
		{
			GetApplication().VersionOtb = 447;
			int result = GetApplication().Initialize();
			if (ResultCode.SUCCEEDED(result))
			{
				GetApplication().EnableTracing(					
					EnumTraceGroup.ALL,
					EnumTraceGroup.ALL,
					EnumTraceGroup.CLIENT,
					EnumTraceGroup.CLIENT,
					"Trace.txt",					
					1000000,
					2);
			}	//	end if
			return result;
		}	//	end Initialize

		public int ProcessCommandLine(string commandLine)
		{
			return Application.Instance.ProcessCommandLine(commandLine);
		}	//	end ProcessCommandLine


		public void Terminate()
		{
			GetApplication().Terminate();
			m_daSession = null;
			m_daSubscription = null;
			m_daItem = null;
			m_aeSession = null;
			m_aeSubscription = null;
		}	//	end Terminate


		public string GetConditionState()
		{
			if (m_aeSession == null)
			{
				return "";
			}	//	end if

			string message = String.Empty;

			try
			{
				string conditionStateToString = String.Empty;
				AeConditionState conditionState = null;

				int result = m_aeSession.GetAeConditionState(
								"computer.clock.time slot 1",
								"between",
								null,
								out conditionState,
								null);
				
				if (ResultCode.SUCCEEDED(result)){

					message = "The condition state is: \n";
					message += " Source Path: computer.clock.time slot 1 \n";
					message += " Condition Name: between \n";
					message += " State: " + GetStateToString(conditionState.State)+"\n";
					message += " Quality: " + conditionState.Quality +"\n";
					message += " Active Time: " + conditionState.ConditionActiveTime +"\n";
					message += " Inactive Time: " + conditionState.ConditionInactiveTime +"\n";
					message += " AcknowledgeTime: " + conditionState.ConditionAckTime +"\n";
					message += " AcknowledgerComment: " + conditionState.AcknowledgerComment +"\n";
					message += " AcknowledgerID: " + conditionState.AcknowledgerId +"\n";
					message += " Active subcondition time: " + conditionState.SubConditionActiveTime +"\n";
					message += " Active subcondition name: " + conditionState.ActiveSubConditionName +"\n";
					message += " Active subcondition definition: " + conditionState.ActiveSubConditionDefinition +"\n";
					message += " Active subcondition description: " + conditionState.ActiveSubConditionDescription +"\n";
					message += " Active subcondition severity: " + conditionState.ActiveSubConditionSeverity +"\n";
					message += " Number of subconditions: " + conditionState.SubConditionsNames.Length +"\n";
					for (int i = 0; i < conditionState.SubConditionsNames.Length; i++)
					{
						message += "	Subcondition name: " + conditionState.SubConditionsNames[i]+"\n";
						message += "	Subcondition definition: " + conditionState.SubConditionsDefinitions[i]+"\n";
						message += "	Subcondition description: " + conditionState.SubConditionsDescriptions[i]+"\n";
						message += "	Subcondition severity: " + conditionState.SubConditionsSeverities[i]+"\n";
					}//end for
				}
				else
				{
					message = "Get condition state failed!\n";
				}	//	end if...else
			}
			catch (Exception exc)
			{
				GetApplication().Trace(
					EnumTraceLevel.ERR,
					EnumTraceGroup.USER,
					"OpcClient::GetConditionState",
					exc.ToString());
			}	//	end try...catch
			return message;
		}//end GetConditionState

		public int InitializeAeObjects()
		{
			int connectResult = (int)EnumResultCode.E_FAIL;

			try{
				
				m_aeSession = new AeSession("opcae:///Softing.OPCToolboxDemo_ServerAE.1/{2E565243-B238-11D3-842D-0008C779D775}");
				m_aeSubscription = new AeSubscription(m_aeSession);				

				connectResult = m_aeSession.Connect(
					true,
					true,
					null);

				m_aeSubscription.RefreshAeConditions(null);
			}
			catch(Exception exc)
			{
				GetApplication().Trace(
					EnumTraceLevel.ERR,
					EnumTraceGroup.USER,
					"OpcClient::InitializeAeObjects",
					exc.ToString());
			}	//	end try...catch

			return connectResult;
		}	//	end InitializeAeObjects
		public int InitializeDaObjects()
		{
			int connectResult = (int)EnumResultCode.E_FAIL;

			try{
				m_daSession = new DaSession("opcda:///Softing.OPCToolboxDemo_ServerDA.1/{2E565242-B238-11D3-842D-0008C779D775}");
				m_daSubscription = new DaSubscription(1000, m_daSession);
				m_daItem = new DaItem("maths.sin", m_daSubscription);

				connectResult = m_daSession.Connect(
					true,
					false,
					null);
			}
			catch(Exception exc)
			{
				GetApplication().Trace(
					EnumTraceLevel.ERR,
					EnumTraceGroup.USER,
					"OpcClient::InitializaDaObjects",
					exc.ToString());
			}	//	end try...catch

			return connectResult;
		}	//	end InitializeDaObjects

		public string ReadItem()
		{
			string message = String.Empty;

			try
			{
				uint readCount = 1;
				DaItem[] itemList = new DaItem[readCount];
				itemList[0] = m_daItem;

				ValueQT[] values = null;
				int[] results = null;

				if (ResultCode.SUCCEEDED(
					m_daSubscription.Read(
						0,
						itemList,
						out values,
						out results,
						null)))
				{
					message += " \nRead item synchronously using subscription \n";

					for (int i = 0; i< values.Length;i++)
					{
						if (ResultCode.SUCCEEDED(results[i]))
						{
							message += " " + itemList[i].Id + " - ";
							message += values[i].ToString() + "\n\n";
						}
						else{
							message += "Read failed for item " + itemList[i].Id + "\n\n";
						}	//	end if...else
					}	//	end for
				}
				else
				{
					message += " Subscription synchronous read failed!" + "\n\n";
				}	//	end if...else
			}
			catch(Exception exc)
			{
				GetApplication().Trace(
					EnumTraceLevel.ERR,
					EnumTraceGroup.USER,
					"OpcClient::ReadItem",
					exc.ToString());
			}	//	end try...catch
			return message;
		}	//	end ReadItem


	  public void Trace(
		  EnumTraceLevel traceLevel,
		  EnumTraceGroup traceGroup,
		  string objectID,
		  string message) 
	  {
		  Application.Instance.Trace(
			  traceLevel,
			  traceGroup,
			  objectID,
			  message);
	  }	//	end Trace

		//--
		#endregion
		
		#region Public Properties
		//-----------------------

		public string ServiceName
		{
			get
			{
				return Application.Instance.ServiceName;
			}
			set
			{
				Application.Instance.ServiceName = value;
			}
		}	//	end SetServiceName

		//--
		#endregion

	}	//	end class OpcClient

}	//	end namespace
